File: packages/plugin-sdk/src/process.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

export interface ProcessCommand {
  /** Absolute executable path. argv is never interpreted by a shell. */
  executable: string;
  args?: readonly string[];
  env?: Readonly<Record<string, string>>;
  /** Default: invocation. Instance processes survive calls, not plugin retirement. */
  lifetime?: 'invocation' | 'instance';
}
export interface ProcessExit {
  code: number | null;
  success: boolean;
  stopped: boolean;
  error: string | null;
}
export interface ProcessChunk {
  stream: 'stdout' | 'stderr';
  bytes: Uint8Array;
}
export interface Process {
  readonly id: string;
  /** Strings are UTF-8; bytes are unchanged. Resolves after bounded queue admission. */
  write(data: string | Uint8Array | readonly number[]): Promise<void>;
  endInput(): Promise<void>;
  /** One pending read. null means end of output, not successful process exit. */
  next(): Promise<ProcessChunk | null>;
  /** Also waits for owned descendants and output cleanup. */
  wait(): Promise<ProcessExit>;
  close(): Promise<void>;
}
export interface Processes {
  /** Requires this call's current Bypass permission and uses its frozen workspace. */
  spawn(command: ProcessCommand): Promise<Process>;
  /** Rebinds an instance process to this invocation; old call handles stay revoked. */
  open(id: string): Process;
}

File: crates/plugins/src/process.rs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

use crate::call::Scope;
use futures_util::future::BoxFuture;
use serde::{Deserialize, Serialize};
use std::{collections::BTreeMap, sync::Arc};

#[derive(Clone, Copy, Default, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum Lifetime {
    #[default]
    Invocation,
    Instance,
}
#[derive(Clone, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Command {
    pub executable: String,
    #[serde(default)]
    pub args: Vec<String>,
    #[serde(default)]
    pub env: BTreeMap<String, String>,
    #[serde(default)]
    pub lifetime: Lifetime,
}
impl Command {
    pub fn validate(&self) -> Result<(), Error> {
        if !std::path::Path::new(&self.executable).is_absolute()
            || self.args.len() > 256
            || self.env.len() > 128
            || serde_json::to_vec(&(&self.executable, &self.args, &self.env))
                .map_err(|e| Error::Invalid(e.to_string()))?
                .len()
                > 64 * 1024
            || self.executable.contains('\0')
            || self.args.iter().any(|arg| arg.contains('\0'))
            || self.env.iter().any(|(key, value)| {
                key.is_empty() || key.contains(['=', '\0']) || value.contains('\0')
            })
        {
            return Err(Error::Invalid(
                "invalid process command or launch limit exceeded".into(),
            ));
        }
        Ok(())
    }
}

/// A lookup key in this activation, never an execution permission or durable PID.
#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
#[serde(transparent)]
pub struct Id(pub String);

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Exit {
    pub code: Option<i32>,
    pub success: bool,
    pub stopped: bool,
    pub error: Option<String>,
}
#[derive(Clone, Copy, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum Stream {
    Stdout,
    Stderr,
}
#[derive(Serialize)]
pub struct Chunk {
    pub stream: Stream,
    pub bytes: Vec<u8>,
}

pub struct Handle {
    pub id: Id,
    pub io: Arc<dyn Process>,
}
pub trait Processes: Send + Sync {
    fn spawn(&self, call: Scope, command: Command) -> BoxFuture<'_, Result<Handle, Error>>;
    /// Bind an instance-owned process to a new authorized call. Old bindings
    /// remain expired; the ID alone cannot keep their permissions alive.
    fn open(&self, call: Scope, id: Id) -> Result<Handle, Error>;
    /// Cleanup by owner remains allowed after the call binding expires.
    fn close(&self, id: Id) -> BoxFuture<'_, Result<(), Error>>;
}
/// Fiber owns the process, not this view. Dropping a view does not terminate an
/// instance-owned process; close explicitly or let Fiber retirement clean it up.
pub trait Process: Send + Sync {
    fn write(&self, bytes: Vec<u8>) -> BoxFuture<'_, Result<(), Error>>;
    fn end_input(&self) -> BoxFuture<'_, Result<(), Error>>;
    fn next(&self) -> BoxFuture<'_, Result<Option<Chunk>, Error>>;
    fn wait(&self) -> BoxFuture<'_, Result<Exit, Error>>;
    fn close(&self) -> BoxFuture<'_, Result<(), Error>>;
}
#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error("process authority is retired or denied")]
    Denied,
    #[error("invalid process operation: {0}")]
    Invalid(String),
    #[error("process operation failed: {0}")]
    Failed(String),
    #[error("process cleanup is unconfirmed: {0}")]
    CleanupUnconfirmed(String),
}

impl From<crate::execution::CommandError> for Error {
    fn from(error: crate::execution::CommandError) -> Self {
        use crate::execution::CommandError;
        match error {
            CommandError::Denied | CommandError::Revoked => Self::Denied,
            CommandError::Invalid(reason) => Self::Invalid(reason),
            CommandError::OutcomeUnknown(reason) => Self::CleanupUnconfirmed(reason),
            error => Self::Failed(error.to_string()),
        }
    }
}
